presto-main/src/main/java/com/facebook/presto/util/KerberosPrincipal.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class KerberosPrincipal
+{
+    private static final char REALM_SEPARATOR = '@';
+    private static final char PARTS_SEPARATOR = '/';
+    private static final char ESCAPE_CHARACTER = '\\';
+
+    private static final String HOSTNAME_PLACEHOLDER = "_HOST";
+
+    private final String username;
+    private final Optional<String> hostname;
+    private final Optional<String> realm;
```

Comment:
put this below static methods. `Parser` class would be helpful here, as it would make this code more readable.